runtime/casts.cu

```diff
@@ -6,6 +6,238 @@
  */
 // clang-format on
 
+// clang-format off
+// Disable clang-format because it tries to put the _Pragma("unroll")
+// and the for loop on the same line, which doesn't make sense.
+#define DEFINE_CAST_VECN_WITH_VEC2(name, from_type, to_type)        \
+  template <int n, int align>                                       \
+  __device__ __inline__ Array<to_type, n, align> name(              \
+      const Array<from_type, n, align>& input) {                    \
+    Array<to_type, n, align> result;                                \
+    _Pragma("unroll")                                               \
+    for (int i = 0; i < n; i += 2) {                                \
+      if (i + 1 < n) {                                              \
+        Array<from_type, 2, align> pair = {input[i], input[i + 1]}; \
+        Array<to_type, 2, align> res_pair = name(pair);             \
+        result[i] = res_pair[0];                                    \
+        result[i + 1] = res_pair[1];                                \
+      } else {                                                      \
+        result[i] = name(input[i]);                                 \
+      }                                                             \
+    }                                                               \
+    return result;                                                  \
+  }
+
+#define DEFINE_CAST_NAIVE_VECN(name, from_type, to_type) \
+  template <int n, int align>                            \
+  __device__ __inline__ Array<to_type, n, align> name(   \
+      const Array<from_type, n, align>& input) {         \
+    Array<to_type, n, align> result;                     \
+    _Pragma("unroll")                                    \
+    for (int i = 0; i < n; i ++) {                       \
+      result[i] = name(input[i]);                        \
+    }                                                    \
+    return result;                                       \
+  }
+
+#define DEFINE_CAST_NAIVE_VECN_FROM_TEMPLATE(name, to_type) \
+  template <typename T, int n, int align>                   \
+  __device__ __inline__ Array<to_type, n, align> name(      \
+      const Array<T, n, align>& input) {                    \
+    Array<to_type, n, align> result;                        \
+    _Pragma("unroll")                                       \
+    for (int i = 0; i < n; i++) {                           \
+      result[i] = name(input[i]);                           \
+    }                                                       \
+    return result;                                          \
+  }
+// clang-format on
+
+// Wrappers for static casts
+template <typename T>
+__device__ __inline__ float __to_float(const T f) {
```

Comment:
This is used to replace `(float)f` in casting expr, because `__to_float` can be easily modified to handle array.